jcl/src/java.base/share/classes/openj9/internal/foreign/abi/InternalDowncallHandler.java

```diff
@@ -323,18 +327,21 @@ private final long memSegmtOfPtrToLongArg(MemorySegment argValue, LinkerOptions
 			 * is captured so as to reset the internal index and avoid retaining the references
 			 * to the unreachable objects.
 			 */
-			if (info != null) {
-				info.clear();
+			if(!flagCreateHeapArgInfo.get() && !infoStack.isEmpty()) {
+				infoStack.pop();
+				flagCreateHeapArgInfo.set(Boolean.TRUE);
 			}
 			/*[ENDIF] JAVA_SPEC_VERSION >= 22 */
 			throw e;
 		}
 
 		long address = argValue.address();
 		/*[IF JAVA_SPEC_VERSION >= 22]*/
-		if (info == null) {
+
+		if ((info == null) || flagCreateHeapArgInfo.get()) {
```

Comment:
The `(info == null)` check is not necessary since `flagCreateHeapArgInfo` is true by default.